# Lab: Building a Reproducible Analysis Workflow

In the first part of this lab, you will use official National Park Service data to answer a guided question. In the second part, you will find a dataset and conduct an analysis of your own.

Use the workflow from lecture throughout the lab:

1. **Question**
2. **Data**
3. **Operation**
4. **Check**
5. **Evidence**
6. **Conclusion**
7. **Limitation**


## Learning goals

By the end of this lab, you should be able to:

- organize an investigation with the reproducible analysis workflow;
- inspect a dataset before analyzing it;
- group, summarize, check, and sort observations with Pandas;
- formulate a research question that can be answered with data; and
- distinguish evidence, conclusions, and limitations.


## Using AI tools responsibly

You may use AI to ask questions, debug code, or receive feedback. You remain responsible for understanding your work, verifying suggestions against the data, and citing the source of your dataset. At the end, disclose whether you used an AI tool. Using AI is not required.


# Part 1: Florida park visitation


## Background

The National Park Service (NPS) publishes estimates of recreational visits to its units. These estimates help parks understand patterns in public use and plan staffing, maintenance, transportation, and visitor services.

This course file contains monthly estimates for five Florida NPS units from 2015 through 2025:

- Big Cypress National Preserve
- Biscayne National Park
- Canaveral National Seashore
- Dry Tortugas National Park
- Everglades National Park

A recreational visit is a visit, not necessarily one unique person. A person who visits several times may be counted several times.


| Feature | Description |
| --- | --- |
| `park_code` | Four-letter NPS unit code |
| `park` | Name of the NPS unit |
| `year` | Calendar year of the observation |
| `month_number` | Month represented as an integer from 1 through 12 |
| `month` | Month name |
| `recreation_visits` | Estimated recreational visits during that park-month |


## Step 1: Question

**After the decline in July 2020, when did Canaveral National Seashore's July recreational visitation first return to or exceed its average July visitation from 2015–2019?**

The 2015–2019 mean provides a pre-pandemic comparison point. In this question, “recovery” means the first July from 2020 onward with visitation at or above that baseline. This definition does not require every later July to remain above the baseline.


## Step 2: Data

Import Pandas using the alias `pd`. Read `data/nps_florida_park_visitation_2015_2025.csv` into a DataFrame named `park_visits`, then display its first five rows.


In [2]:
# Write your code here.
import pandas as pd
park_visits = pd.read_csv('data/nps_florida_park_visitation_2015_2025.csv')
park_visits.head()

,park_code,park,year,month_number,month,recreation_visits
0,BICY,Big Cypress National Preserve,2015,1,January,114045
1,BICY,Big Cypress National Preserve,2015,2,February,144112
2,BICY,Big Cypress National Preserve,2015,3,March,153084
3,BICY,Big Cypress National Preserve,2015,4,April,113880
4,BICY,Big Cypress National Preserve,2015,5,May,72510


Display the last five rows.


In [3]:
# Write your code here.
park_visits.tail()

,park_code,park,year,month_number,month,recreation_visits
655,EVER,Everglades National Park,2025,8,August,50464
656,EVER,Everglades National Park,2025,9,September,32290
657,EVER,Everglades National Park,2025,10,October,19008
658,EVER,Everglades National Park,2025,11,November,53416
659,EVER,Everglades National Park,2025,12,December,70926


Display information about the DataFrame and calculate the number of missing observations in each feature.


In [4]:
# Write your code here.
park_visits.info()
park_visits.isnull().sum()

<class 'pandas.DataFrame'>
RangeIndex: 660 entries, 0 to 659
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   park_code          660 non-null    str  
 1   park               660 non-null    str  
 2   year               660 non-null    int64
 3   month_number       660 non-null    int64
 4   month              660 non-null    str  
 5   recreation_visits  660 non-null    int64
dtypes: int64(3), str(3)
memory usage: 54.1 KB


park_code            0
park                 0
year                 0
month_number         0
month                0
recreation_visits    0
dtype: int64

**What does each row represent?**

Your answer: One parks visit count for a specific month in a specific year.


**Are any observations missing from the six features?**

Your answer: No


## Step 3: Operation

Create one Boolean condition that identifies Canaveral National Seashore and another that identifies July.


In [5]:
# Write your code here.
is_canaveral = park_visits["park"] == "Canaveral National Seashore"
is_july = park_visits["month"] == "July"

Combine the two conditions with `&`. Use `.loc` to select the `year` and `recreation_visits` features, make a copy named `canaveral_july`, and display it.


In [7]:
# Write your code here.
combinedCond = is_canaveral & is_july
canaveral_july = park_visits.loc[combinedCond, ["year", "recreation_visits"]].copy()
canaveral_july

,year,recreation_visits
270,2015,163169
282,2016,216437
294,2017,138018
306,2018,189285
318,2019,162947
330,2020,112898
342,2021,206636
354,2022,233352
366,2023,173296
378,2024,176242


Create `pre_pandemic_condition`, which is `True` for years from 2015 through 2019. Use it with `.loc` to calculate the mean of `recreation_visits`, and store the result as `pre_pandemic_july_mean`.

Display `pre_pandemic_july_mean`.


In [36]:
# Write your code here.
pre_pandemic_condition = (park_visits["year"] >= 2015) & (park_visits["year"] <= 2019)
pre_pandemic_july_mean = park_visits.loc[combinedCond & pre_pandemic_condition, "recreation_visits"].mean()
print(pre_pandemic_july_mean)

173971.2


Create a condition that identifies rows from 2020 onward whose `recreation_visits` are greater than or equal to `pre_pandemic_july_mean`. Use `.loc` to create `recovery_years`, then display it in ascending year order.


In [32]:
# Write your code here.
recovCond = is_canaveral & is_july & (park_visits["year"] >= 2020) & (park_visits["recreation_visits"] >= pre_pandemic_july_mean)
recovery_years = park_visits.loc[recovCond].sort_values("year")
print(recovery_years)

    park_code                         park  year  month_number month  \
330      CANA  Canaveral National Seashore  2020             7  July   
342      CANA  Canaveral National Seashore  2021             7  July   
354      CANA  Canaveral National Seashore  2022             7  July   
366      CANA  Canaveral National Seashore  2023             7  July   
378      CANA  Canaveral National Seashore  2024             7  July   
390      CANA  Canaveral National Seashore  2025             7  July   

     recreation_visits  
330             112898  
342             206636  
354             233352  
366             173296  
378             176242  
390             253403  


## Step 4: Check

Display the number of rows in `canaveral_july`, its smallest year, and its greatest year.


In [27]:
# Write your code here.
print("Number of rows: ", len(canaveral_july))
print("Smallest year: " , canaveral_july["year"].min())
print("Greatest year: " , canaveral_july["year"].max())

Number of rows:  11
Smallest year:  2015
Greatest year:  2025


**Do these results confirm that the filtered data contains one July observation for every year from 2015 through 2025?**

Your answer: Yes, 11 years counting 2015.


Count the observations used to calculate `pre_pandemic_july_mean`.


In [39]:
# Write your code here.
pre_pandemic_count = park_visits.loc[combinedCond & pre_pandemic_condition, "recreation_visits"].count()
print(pre_pandemic_count)

5


**Does the baseline use the expected five years?**

Your answer: yes. pre_pandemic_count returned 5. So 5 observed July's


## Step 5: Evidence

Display the first row of `recovery_years`.


In [40]:
# Write your code here.
recovery_years.head(1)

,park_code,park,year,month_number,month,recreation_visits
330,CANA,Canaveral National Seashore,2020,7,July,112898


**What was the 2015–2019 mean, and which year first met or exceeded it after the July 2020 decline? Report that year's July visitation.**

Your answer: Mean = 173971.2
In 2021 it exceeded the mean at 206636 visitations

## Step 6: Conclusion


**Write one sentence that answers the research question using the evidence.**

Your answer: Canaveral National Seashore exceeded its average July visitation from 2015-2019 in 2021 with 206,636 visits.


## Step 7: Limitation


**Does first exceeding the baseline mean that every later July remained above it? Use the data to explain.**

Your answer:No, in July 2023. It actually fell below the baseline at 173296 visits.


**Why does this analysis alone not establish that the pandemic caused the 2020 decline?**

Your answer:


# Part 2: Your own investigation


## Find a dataset

Find a CSV dataset that interests you. Kaggle is one possible source, but you may use another reputable source. Choose a dataset that:

- has a clearly identified publisher or creator;
- includes documentation explaining the observations and features;
- contains at least one categorical and one quantitative feature;
- is appropriate to share in a class assignment;
- does not contain private or personally identifying information; and
- is small enough to run comfortably in this notebook.

Save the CSV in the notebook's `data` folder before continuing.


**What is the dataset's title, publisher or creator, and direct source URL?**

Your answer:https://www.kaggle.com/datasets/mattop/best-selling-nintendo-64-video-games?resource=download
Matt OP
Best Selling Nintendo 64 Video Games

**What does one row represent?**

Your answer: The game title, Developer, Publisher, Release Date, and units sold.


## Step 1: Question

Formulate one research question that can be answered with your dataset. The question must require a calculation or comparison, not merely looking up one value.


**What is your research question?**

Your answer: Did first-party Nintendo Developed games outperform third-party games in terms of units sold?


## Step 2: Data

Read your CSV into a clearly named DataFrame. Display the first and last five rows, information about the DataFrame, and the missing-value count for every feature.


In [48]:
# Write your code here.
n64Games = pd.read_csv('data/best-selling-nintendo64.csv')
print("HEAD:\n", n64Games.head())
print("TAIL:\n" ,n64Games.tail())
n64Games.info()
n64Games.isnull().sum()

HEAD:
                                    Game    Developer(s) Publisher(s)  \
0                        Super Mario 64    Nintendo EAD     Nintendo   
1                         Mario Kart 64    Nintendo EAD     Nintendo   
2                         GoldenEye 007            Rare     Nintendo   
3  The Legend of Zelda: Ocarina of Time    Nintendo EAD     Nintendo   
4                     Super Smash Bros.  HAL Laboratory     Nintendo   

  Release date     Sales  
0   1996-06-23  11910000  
1   1996-12-14   9870000  
2   1997-08-25   8090000  
3   1998-11-21   7600000  
4   1999-01-21   5550000  
TAIL:
                             Game  \
41                 Pilotwings 64   
42                      F-Zero X   
43      Pocket Monsters' Stadium   
44  007: The World Is Not Enough   
45               Namco Museum 64   

                                         Developer(s)     Publisher(s)  \
41  Nintendo EAD / Nintendo R&D3 / Paradigm Entert...         Nintendo   
42                        

Game            0
Developer(s)    0
Publisher(s)    0
Release date    0
Sales           0
dtype: int64

**Which observations and features will you use to answer your question?**

Your answer: Developers, and Sales (Units Sold)


## Step 3: Operation

Write the code needed to prepare, filter, group, summarize, or sort the data. Store the final evidence in a clearly named object and display it.


In [62]:
# Write your code here.
#check firstparty
n64Games['FirstParty'] = n64Games['Developer(s)'].str.contains('Nintendo EAD|HAL Laboratory', case=False, regex=True)
#Group and summarize
sales_summary = n64Games.groupby('FirstParty')['Sales'].agg(['count', 'mean', 'median', 'sum'])
#Renaming for readability
sales_summary = sales_summary.rename(columns={'count': 'Titles'})
sales_summary = sales_summary.rename(columns={'sum': 'Units Sold'})
#Sort
sales_summary = sales_summary.sort_values('mean', ascending=False)

#formating for readability
pd.set_option('display.float_format', '{:.0f}'.format)
final = sales_summary
final

,Titles,mean,median,Units Sold
FirstParty,,,,
True,16,4176548,3150000,66824765
False,30,2305667,1855000,69170000


**Explain how your operations connect the selected data to your research question.**

Your answer: My operations divided the data into two. First party developed games (Nintendo and HAL) and grouped first party, and 3rd party developers together. 
Then calculated there Titles, Mean, Median, and Total sold Units. 


## Step 4: Check

Write at least one check that could reveal an error in your analysis, then display its result.


In [61]:
# Write your code here.
check = n64Games[n64Games['Developer(s)'].str.contains('Nintendo EAD|HAL Laboratory', case=False, regex=True) & (n64Games['FirstParty'] == False)]
check

,Game,Developer(s),Publisher(s),Release date,Sales,FirstParty


**What did you check, and did the result meet your expectation?**

Your answer: I checked to be sure that no title that contained Either Nintendo EAD, or HAL Laboratory was not grouped into the 3rd party by mistake.
The result returned no errors which I expected.


## Step 5: Evidence


**Which specific value or comparison in your output answers the question?**

Your answer: Sales (Units Sold).


## Step 6: Conclusion


**What conclusion is supported by the evidence?**

Your answer: That 3rd party games actually outperformed 1st party games in terms of sales by about 3 million according to this data set.


## Step 7: Limitation


**What should a reader avoid concluding from your analysis, and why?**

Your answer: That 3rd party games on this system were actually better. Because in this data set, the amount of 3rd party games is nearly double the amount of 1st party games.


## AI-use reflection


**State whether you used an AI tool. If you did, describe one suggestion you checked and what you accepted, changed, or rejected. If you did not, state that you did not use one.**

Your response: I used Copilot to assist me with renaming the columns in sales_summary and to adjust the mean using pandas, as it was being displaying in scientific notation. The first suggestion it gave me when renaming the columns used lambda. I am unfamiliar with it, and have not used
it in python. So I rejected that suggestion and 


## Submission checklist

Before submitting, confirm that all cells run in order, the independent dataset and its source are included, every answer is complete, numerical claims match the output, conclusions are appropriately limited, and the AI-use reflection is complete.
